### PR TITLE
DefaultPromise concurrency bug

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
If the executor changes while listeners are added and notification of listeners is being done then listeners can be notified out of order and concurrently. We should ensure that only one executor is used at any given time to notify listeners and ensure the listeners are notified in FIFO order.

Modifications:
- Move the notifyingListeners member variable from DefaultPromise into the synchronized block to prevent concurrent notification of listeners and preserve FIFO notification order

Result:
If the executor is changed for a DefaultPromise the listener notification order should be FIFO.

### CODE REVIEW

**Comment by Scottmitch** [comment]
@ejona86 - PTAL

**Comment by normanmaurer** [comment]
LGTM... will wait for @ejona86 to review tho.

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java [hunk 1]
```diff
@@ -431,24 +432,27 @@ public void run() {
     }
 
     private void notifyListeners0() {
-        Object listeners;
-        while (!notifyingListeners) {
+        Object listeners = null;
+        for (;;) {
             synchronized (this) {
-                if (this.listeners == null) {
+                // Only proceed if there are listeners to notify and we are not already notifying listeners.
```
Comment (ejona86) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java]: This looks correct but is a bit hard to follow. Maybe special-case the first iteration? More lines of code but seems more readable and less prone to error.

```java
Object listeners;
synchronized (this) {
  if (notifyingListeners || this.listeners == null) {
    return;
  }
  notifyingListeners = true;
  listeners = this.listeners;
  this.listeners = null;
}
for (;;) {
  blahNotifyListener0Stuff();
  synchronized (this) {
    if (this.listeners == null) {
      notifyingListeners = false;
      return;
    }
    listeners = this.listeners;
    this.listeners = null;
  }
}
```

**Comment by ejona86** [comment]
LGTM

File: common/src/main/java/io/netty/util/concurrent/DefaultPromise.java (hunk 1 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/concurrent/DefaultPromise.java]: sure I can adjust the code to this approach

**Comment by Scottmitch** [comment]
Cherry-picked 4.1 (f5d58e2) 4.0 (c9a29c4)

**Comment by Scottmitch** [comment]
thanks for review @ejona86 !
